internal/command/test.go

```diff
@@ -387,22 +537,148 @@ type TestModuleState struct {
+// variable values that are relevant to the config being tested.
+//
+// Crucially, it differs from buildInputVariablesForAssertions in that it only
+// includes variables that are reference by the config and not everything that
+// is defined within the test run block and test file.
+func buildInputVariablesForTest(run *moduletest.Run, file *moduletest.File, config *configs.Config) (terraform.InputValues, tfdiags.Diagnostics) {
+	variables := make(map[string]hcl.Expression)
+	for name := range config.Module.Variables {
+		if run != nil {
+			if expr, exists := run.Config.Variables[name]; exists {
+				// Local variables take precedence.
+				variables[name] = expr
+				continue
+			}
+		}
+
+		if file != nil {
+			if expr, exists := file.Config.Variables[name]; exists {
+				// If it's not set locally, it maybe set globally.
+				variables[name] = expr
+				continue
+			}
+		}
+
+		// If it's not set at all that might be okay if the variable is optional
+		// so we'll just not add anything to the map.
+	}
+
+	unparsed := make(map[string]backend.UnparsedVariableValue)
+	for key, value := range variables {
+		unparsed[key] = unparsedVariableValueExpression{
+			expr:       value,
+			sourceType: terraform.ValueFromConfig,
+		}
+	}
+	return backend.ParseVariableValues(unparsed, config.Module.Variables)
+}
+
+// buildInputVariablesForAssertions creates a terraform.InputValues mapping that
+// contains all the variables defined for a given run and file, alongside any
+// unset variables that have defaults within the provided config.
+//
+// Crucially, it differs from buildInputVariablesForTest in that the returned
+// input values include all variables available even if they are not defined
+// within the config.
+//
+// This does mean the returned diags might contain warnings about variables not
+// defined within the config. We might want to remove these warnings in the
+// future, since it is actually okay for test files to have variables defined
+// outside the configuration.
+func buildInputVariablesForAssertions(run *moduletest.Run, file *moduletest.File, config *configs.Config) (terraform.InputValues, tfdiags.Diagnostics) {
+	merged := make(map[string]hcl.Expression)
+
+	if run != nil {
+		for name, expr := range run.Config.Variables {
+			merged[name] = expr
+		}
+	}
+
+	if file != nil {
+		for name, expr := range file.Config.Variables {
+			if _, exists := merged[name]; exists {
+				// Then this variable was defined at the run level and we want
+				// that value to take precedence.
+				continue
+			}
+			merged[name] = expr
+		}
+	}
+
+	unparsed := make(map[string]backend.UnparsedVariableValue)
+	for key, value := range merged {
+		unparsed[key] = unparsedVariableValueExpression{
+			expr:       value,
+			sourceType: terraform.ValueFromConfig,
+		}
+	}
+	return backend.ParseVariableValues(unparsed, config.Module.Variables)
+}
+
+func captureTestPanic(identifier string, diags *tfdiags.Diagnostics) {
```

Comment:
If we are trying to handle panics in the tests, we may want to look into overriding the behavior of the global panic handler. Most codepaths in core are covered by `logging.PanicHandler`, which does not recover and exits with 11 after writing the custom error message. (this also might be a rare enough occurrence that it's not really worth the effort)

Reply:
ah yeah, I see that `logging.PanicHandler` is also referenced directly in the graph walk cycle. Let me take this functionality out of this PR, and I can make a dedicated PR separately that can solve this. It's not as simple as I thought it would be.